db/kvstorewithcache.go

```diff
@@ -0,0 +1,150 @@
+func (kvc *kvStoreWithCache) Get(namespace string, key []byte) ([]byte, error) {
+	cachedData, err := kvc.getStateCaches(namespace, key)
+	if err != nil {
+		return nil, err
+	}
+	if cachedData != nil {
+		return cachedData, nil
+	}
+	kvStoreData, err := kvc.store.Get(namespace, key)
+	if err != nil {
+		return nil, err
+	}
+	// in case of read-miss, put into statecaches
+	kvc.putStateCaches(namespace, key, kvStoreData)
+	return kvStoreData, nil
+}
+
+// Filter returns <k, v> pair in a bucket that meet the condition
+func (kvc *kvStoreWithCache) Filter(namespace string, cond Condition, minKey, maxKey []byte) ([][]byte, [][]byte, error) {
+	return kvc.store.Filter(namespace, cond, minKey, maxKey)
+}
+
+// Delete deletes a record from statecaches if exists, and from kvstore
+func (kvc *kvStoreWithCache) Delete(namespace string, key []byte) (err error) {
+	kvc.deleteStateCachesIfExists(namespace, key)
+	return kvc.store.Delete(namespace, key)
+}
+
+// WriteBatch commits a batch into stateCaches and kvstore
+func (kvc *kvStoreWithCache) WriteBatch(kvsb batch.KVStoreBatch) (err error) {
+	for i := 0; i < kvsb.Size(); i++ {
+		write, e := kvsb.Entry(i)
+		if e != nil {
+			return e
+		}
+		ns := write.Namespace()
+		if write.WriteType() == batch.Put {
+			kvc.updateStateCachesIfExists(ns, write.Key(), write.Value())
+		} else if write.WriteType() == batch.Delete {
+			kvc.deleteStateCachesIfExists(ns, write.Key())
+		}
+	}
+	return kvc.store.WriteBatch(kvsb)
+}
+
+// ======================================
+// private functions
+// ======================================
+
+// store on stateCaches
+func (kvc *kvStoreWithCache) putStateCaches(namespace string, key, value []byte) {
+	kvc.mutex.Lock()
+	defer kvc.mutex.Unlock()
+	if sc, ok := kvc.stateCaches[namespace]; ok {
+		sc.Add(string(key), value)
+	} else {
+		sc := cache.NewThreadSafeLruCache(kvc.cacheSize)
+		sc.Add(string(key), value)
+		kvc.stateCaches[namespace] = sc
+	}
+	return
+}
+
+// update on stateCaches if the key exists
+func (kvc *kvStoreWithCache) updateStateCachesIfExists(namespace string, key, value []byte) {
+	kvc.mutex.Lock()
+	defer kvc.mutex.Unlock()
+	if sc, ok := kvc.stateCaches[namespace]; ok {
+		if _, ok := sc.Get(string(key)); ok {
+			sc.Add(string(key), value)
+		}
+	}
+	return
+}
+
+// remove on stateCaches if the key exists
+func (kvc *kvStoreWithCache) deleteStateCachesIfExists(namespace string, key []byte) {
+	kvc.mutex.Lock()
+	defer kvc.mutex.Unlock()
+	if sc, ok := kvc.stateCaches[namespace]; ok {
+		sc.Remove(string(key))
+	}
+	return
+}
+
+// look up stateCachces
+func (kvc *kvStoreWithCache) getStateCaches(namespace string, key []byte) ([]byte, error) {
+	kvc.mutex.RLock()
+	defer kvc.mutex.RUnlock()
+	if sc, ok := kvc.stateCaches[namespace]; ok {
+		if data, ok := sc.Get(string(key)); ok {
+			if byteData, ok := data.([]byte); ok {
+				return byteData, nil
+			}
+			return nil, errors.New("failed to convert interface{} to bytes from stateCaches")
```

Comment:
this should be a fatal error, because returning the error won't have the problem fixed in any way. Thus, the return type should be "(value []byte, exist bool)".

Reply:
yup modified on the latest commit!